### PR TITLE
Added Monad3/Error3 cats/zio like syntax enhancements

### PR DESCRIPTION
**Pull request by vitaliihonta**

### CODE REVIEW

File: fundamentals/fundamentals-bio/src/main/scala/izumi/functional/bio/syntax/BIO3Syntax.scala [hunk 1]
```diff
@@ -54,6 +54,12 @@ object BIO3Syntax {
     @inline final def tap[R1 <: R, E1 >: E, B](f0: A => FR[R1, E1, Unit]): FR[R1, E1, A] = F.flatMap[R1, E1, A, A](r)(a => F.map(f0(a))(_ => a))
 
     @inline final def flatten[R1 <: R, E1 >: E, A1](implicit ev: A <:< FR[R1, E1, A1]): FR[R1, E1, A1] = F.flatten(F.widen(r))
+
+    @inline final def iterateWhile(p: A => Boolean): FR[R, E, A] =
```
Comment (neko-kai) [inline fundamentals/fundamentals-bio/src/main/scala/izumi/functional/bio/syntax/BIO3Syntax.scala]: @vitaliihonta Could you please move these changes from `BIO3Syntax` to `Syntax3`? BIO3Syntax is just deprecated aliases for those upgrading from 0.10 series to 1.0

File: fundamentals/fundamentals-bio/src/main/scala/izumi/functional/bio/Error3.scala [hunk 2]
```diff
@@ -27,6 +27,63 @@ trait Error3[F[-_, +_, +_]] extends ApplicativeError3[F] with Monad3[F] {
...
+  def someOrElseM[R, E, A](r: F[R, E, Option[A]])(default: F[R, E, A]): F[R, E, A] =
+    flatMap(r) {
+      case Some(value) => pure(value)
+      case None => default
+    }
+
+  /**
+    * Extracts the optional value, or fails with the given error 'e'.
+    */
+  def someOrFail[R, E, A](r: F[R, E, Option[A]])(e: => E): F[R, E, A] =
```
Comment (neko-kai) [inline fundamentals/fundamentals-bio/src/main/scala/izumi/functional/bio/Error3.scala]: There is already `someOrFail` - it's `ApplicativeError3#fromOption`

File: fundamentals/fundamentals-bio/src/main/scala/izumi/functional/bio/Error3.scala [hunk 3]
```diff
@@ -27,6 +27,63 @@ trait Error3[F[-_, +_, +_]] extends ApplicativeError3[F] with Monad3[F] {
...
+  /**
+    * Extracts the optional value, or returns the given 'default'.
+    */
+  def someOrElse[R, E, A](r: F[R, E, Option[A]])(default: => A): F[R, E, A] =
+    map(r)(_.getOrElse(default))
+
+  /**
+    * Extracts the optional value, or executes the effect 'default'.
+    */
+  def someOrElseM[R, E, A](r: F[R, E, Option[A]])(default: F[R, E, A]): F[R, E, A] =
```
Comment (neko-kai) [inline fundamentals/fundamentals-bio/src/main/scala/izumi/functional/bio/Error3.scala]: Could you please rename this to `fromOptionF` to align with existing `fromOption` method? Also I think it's better to suspend the `default` argument to be consistent with `ApplicativeError3#orElse`

File: fundamentals/fundamentals-bio/src/main/scala/izumi/functional/bio/Error3.scala [hunk 4]
```diff
@@ -27,6 +27,63 @@ trait Error3[F[-_, +_, +_]] extends ApplicativeError3[F] with Monad3[F] {
...
+  /**
+    * Retries this effect while its error satisfies the specified effectful predicate.
+    */
+  def retryWhileM[R, E, A](r: F[R, E, A])(f: E => F[R, Nothing, Boolean]): F[R, E, A] =
+    retryUntilM(r)(e => map(f(e))(!_))
+
+  /**
+    * Extracts the optional value, or returns the given 'default'.
+    */
+  def someOrElse[R, E, A](r: F[R, E, Option[A]])(default: => A): F[R, E, A] =
```
Comment (neko-kai) [inline fundamentals/fundamentals-bio/src/main/scala/izumi/functional/bio/Error3.scala]: I'd like this to be aligned with `fromOption*` naming scheme, but I can't think of a suitable concise name...

File: fundamentals/fundamentals-bio/src/main/scala/izumi/functional/bio/syntax/Syntax2.scala [hunk 5]
```diff
@@ -92,6 +98,27 @@ object Syntax2 {
...
+    @inline final def retryUntilM(f: E => F[Nothing, Boolean]): F[E, A] =
+      F.retryUntilM[Any, E, A](r)(f)
+
+    @inline final def retryWhile(f: E => Boolean): F[E, A] =
+      F.retryWhile(r)(f)
+
+    @inline final def retryWhileM(f: E => F[Nothing, Boolean]): F[E, A] =
+      F.retryWhileM[Any, E, A](r)(f)
+
+    @inline final def someOrElse[B](default: => B)(implicit ev: A <:< Option[B]): F[E, B] =
```
Comment (neko-kai) [inline fundamentals/fundamentals-bio/src/main/scala/izumi/functional/bio/syntax/Syntax2.scala]: This is the same as `fromOption` here - https://github.com/7mind/izumi/pull/1390/files#diff-9018879521a7e6653abbe00676af3367f1b5c0ed3293d54f424fa03e04aa3d8eR99

File: fundamentals/fundamentals-bio/src/main/scala/izumi/functional/bio/Error3.scala [hunk 6]
```diff
@@ -27,6 +27,63 @@ trait Error3[F[-_, +_, +_]] extends ApplicativeError3[F] with Monad3[F] {
...
+  /**
+    * Retries this effect until its error satisfies the specified predicate.
+    */
+  def retryUntil[R, E, A](r: F[R, E, A])(f: E => Boolean): F[R, E, A] =
+    retryUntilM(r)(e => pure(f(e)))
+
+  /**
+    * Retries this effect until its error satisfies the specified effectful predicate.
+    */
+  def retryUntilM[R, E, A](r: F[R, E, A])(f: E => F[R, Nothing, Boolean]): F[R, E, A] =
```
Comment (neko-kai) [inline fundamentals/fundamentals-bio/src/main/scala/izumi/functional/bio/Error3.scala]: Could you please use `F` suffix instead of `M`?

**Review by neko-kai — COMMENTED** [review]
@vitaliihonta Thank you very much! 🙏 I'd say this is good to go except for some naming issues I've detailed in the review comments

File: fundamentals/fundamentals-bio/src/main/scala/izumi/functional/bio/impl/AsyncZio.scala [hunk 7]
```diff
@@ -46,6 +46,12 @@ class AsyncZio extends Async3[ZIO] with Local3[ZIO] {
   @inline override final def <*[R, E, A, B](f: ZIO[R, E, A], next: => ZIO[R, E, B]): ZIO[R, E, A] = f <* next
   @inline override final def map2[R, E, A, B, C](r1: ZIO[R, E, A], r2: => ZIO[R, E, B])(f: (A, B) => C): ZIO[R, E, C] = r1.zipWith(r2)(f)
 
+  @inline override final def iterateWhile[R, E, A](f: ZIO[R, E, A])(p: A => Boolean): ZIO[R, E, A] =
+    f.repeatWhile(p)
```
Comment (neko-kai) [inline fundamentals/fundamentals-bio/src/main/scala/izumi/functional/bio/impl/AsyncZio.scala:50]: Could you add overrides where possible to `AsyncMonix` as well?

File: fundamentals/fundamentals-bio/src/main/scala/izumi/functional/bio/Error3.scala (hunk 4 above)
Comment (vitaliihonta) [reply fundamentals/fundamentals-bio/src/main/scala/izumi/functional/bio/Error3.scala]: @neko-kai What about `fromOptionOr` ? I suppose this name has pretty obvious semantic

**Comment by neko-kai** [comment]
@vitaliihonta Thank you! 🙏

**Pull request merged**